hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataResolver.java

```diff
@@ -56,59 +60,113 @@ public Mapping resolve(String iMapName) {
-        Metadata metadata = hd ? resolveFromHd(container) : resolveFromHeap(iMapName, context);
+        boolean hd = container.getMapConfig().getInMemoryFormat() == InMemoryFormat.NATIVE
+                && !container.getMapConfig().getTieredStoreConfig().isEnabled();
+
+        Metadata metadata = hd
+                ? resolveFromContentsHd(iMapName, context)
+                : resolveFromContents(iMapName, context);
         return metadata == null
                 ? null
                 : new Mapping(iMapName, iMapName, null, TYPE_NAME, null, metadata.fields(), metadata.options());
     }
 
     @Nullable
-    @SuppressWarnings("rawtypes")
-    private Metadata resolveFromHd(MapContainer container) {
-        if (container.getIndexes() == null) {
-            return null;
+    private Metadata resolveFromContents(String name, MapServiceContext context) {
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
+            RecordStore<?> recordStore = partitionContainer.getExistingRecordStore(name);
+            Metadata resolved = resolveFromContentsRecordStore(nodeEngine, recordStore);
+            if (resolved != null) {
+                return resolved;
+            }
         }
+        return null;
+    }
 
-        InternalIndex[] indexes = container.getIndexes().getIndexes();
-        if (indexes == null || indexes.length == 0) {
-            return null;
+    private Metadata resolveFromContentsHd(String name, MapServiceContext context) {
+        // Iterate only over local partitions.
+        // MC will invoke this operation on each member until it finds some data.
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
+            // HD access must be from partition threads.
+            GetAnyMetadataOperation op = new GetAnyMetadataOperation(name);
+            Metadata resolved = invoke(op, partitionContainer.getPartitionId());
+            if (resolved != null) {
+                return resolved;
+            }
         }
+        return null;
+    }
 
-        InternalIndex index = indexes[0];
-        Iterator<QueryableEntry> entryIterator = index.getSqlRecordIterator(false);
-        if (!entryIterator.hasNext()) {
-            return null;
+    private static final class GetAnyMetadataOperation extends Operation
+            implements ReadonlyOperation, PartitionAwareOperation {
+        private final String mapName;
+
+        private GetAnyMetadataOperation(String mapName) {
+            this.mapName = mapName;
+        }
+
+        @Override
+        public void run() {
+            MapService service = getService();
+            MapServiceContext context = service.getMapServiceContext();
+            RecordStore<?> recordStore = context.getExistingRecordStore(getPartitionId(), mapName);
+            // can return null, but that is fine
+            sendResponse(resolveFromContentsRecordStore(getNodeEngine(), recordStore));
```

Comment:
If this `resolveFromContentsRecordStore` method is costly,  maybe operation can return an entry as a response then on caller thread `resolveFromContentsRecordStore` can be called to create metadata. This is to free partition thread quickly.

Reply:
`resolveMetadata` needs access to serialized value (Record). I think it may be safer to do it in partition thread. `resolveMetadata` should not be very costly, it does some lookup for definitions or Java reflection. It is also a rare operation, invoked only by the user from Management Center.